File: frontend_src/node/frontend/src/treasure.ts
import { Map } from "./map";
import { Camera } from "./camera";
import axiosInstance from "./utils/axiosInstance";
import { store } from './store'; // Reduxストアをインポート
import { fetchCurrentUser } from './features/auth/authSlice';

export class Treasure {
  public x: number;
  public y: number;
  private gravity: number;
  private imageClosed: HTMLImageElement;
  private imageOpened: HTMLImageElement;
  private image: HTMLImageElement;
  public rewardImage: HTMLImageElement;
  private width: number;
  private height: number;
  private isOpened: boolean;
  private blinkCounter: number;

  constructor(x: number, y: number, gravity: number, rewardImage: HTMLImageElement) {
    this.x = x;
    this.y = y;
    this.gravity = gravity;
    this.imageClosed = this.loadImage("assets/treasures/treasure_box.png");
    this.imageOpened = this.loadImage("assets/treasures/treasure_box_opened.png");
    this.image = this.imageClosed;
    this.rewardImage = rewardImage;
    this.width = this.imageClosed.width;
    this.height = this.imageClosed.height;
    this.isOpened = false;
    this.blinkCounter = 0;
  }

  private loadImage(src: string): HTMLImageElement {
    const img = new Image();
    img.src = src;
    return img;
  }

  draw(context: CanvasRenderingContext2D, camera: Camera): void {
    if (!this.isOpened || (this.isOpened && this.blinkCounter % 10 < 5)) {
      context.drawImage(this.image, this.x - camera.getX(), this.y - camera.getY());
    }
  }

  draw_on_toppage(context: CanvasRenderingContext2D, x: number, y: number): void {
    this.x = x; // クリック判定のためにx座標を更新
    this.y = y; // クリック判定のためにy座標を更新

    this.width = 500
    this.height = 500
  
    context.save(); // 現在の描画状態を保存
  
    // 描画位置を基準にスケールを反転（-1で反転）
    context.translate(x + this.width / 2, y + this.height / 2); // 中心点に移動
    context.scale(-1, 1); // X軸方向に反転
    context.translate(-(x + this.width / 2), -(y + this.height / 2)); // 元の位置に戻す
  
    if (!this.isOpened) {
      context.drawImage(this.image, x, y, this.width, this.height);
    } else {
      context.drawImage(this.imageOpened, x, y, this.width, this.height);
    }
  
    context.restore(); // 描画状態を元に戻す
  }
    
  update(mapInstance: Map): void {
    const [isOnGround, newY] = mapInstance.isOnGround(this.x, this.y, this.height);
    if (isOnGround) {
      this.y = newY; // 接地している場合、Y座標を修正
    } else {
      this.y += this.gravity; // 重力を適用
    }
  }

  async open(rewardImage: HTMLImageElement, gameMap: Map): Promise<void> {
    if (!this.isOpened) {
      this.isOpened = true;
      this.image = this.imageOpened;
      this.blinkCounter = 30;
  
      // Reduxストアの状態を直接取得
      let currentUser = store.getState().auth.currentUser;
  
      // currentUserがnullの場合、fetchCurrentUserを実行
      if (!currentUser) {
        console.warn("User is not logged in. Attempting to fetch user...");
        const result = await store.dispatch(fetchCurrentUser() as any); // 型アサーションが必要
        if (result.meta.requestStatus === 'fulfilled') {
          currentUser = store.getState().auth.currentUser; // 状態を更新
        } else {
          console.error("Failed to fetch user.");
          return;
        }
      }
  
      // userId の取得時に null チェックを追加
      const userId = currentUser?.id;
      if (!userId) {
        console.error("User ID is missing.");
        return;
      }
  
      // RewardImageからmovie_poster_idを取得
      const moviePosterId = rewardImage.dataset.moviePosterId;
  
      if (moviePosterId) {
        try {
          const response = await axiosInstance.post("/rewards/collect", {
            movie_poster_id: moviePosterId,
            user_id: userId,
          });
  
          if (response.status === 200) {
            console.log(response.data.message);
            gameMap.getCollectedRewards().push(rewardImage); // ここで Map.collectedRewards に追加
          } else {
            console.error("Failed to collect reward");
          }
        } catch (error) {
          console.error("Error:", error);
        }
      }
    }
  }
  
  updateBlink(): void {
    if (this.isOpened && this.blinkCounter > 0) {
      this.blinkCounter--;
    }
  }

  public drawCollectedRewards(
    ctx: CanvasRenderingContext2D,
    collectedRewards: HTMLImageElement[]
  ): void {
    const baseX = 10; // 描画開始位置（X座標）
    const baseY = 10; // 描画開始位置（Y座標）
    const size = 100; // 画像サイズ（幅と高さ）
    const offsetX = 5; // 各アイテム間のX方向のオフセット
    const offsetY = 5;  // 各アイテム間のY方向のオフセット
  
    collectedRewards.forEach((reward, index) => {
      if (reward && reward.complete && reward.naturalWidth > 0) {
        const x = baseX + index * offsetX; // 左方向にずらす
        const y = baseY + index * offsetY; // 下方向にずらす
  
        // 描画
        ctx.drawImage(reward, x, y, size, size);
      } else {
        console.warn(
          `Reward image is not ready to draw or is undefined: ${reward?.src || "undefined"}`
        );
      }
    });
  }

  getRewardImage(){
    return this.rewardImage;
  }
}

File: pygame/app/treasure.py
import pygame
import sys

class Treasure:
    def __init__(self, x, y, gravity, reward_image_path):
        self.x = x
        self.y = y
        self.gravity = gravity
        self.image_closed = pygame.image.load("assets/treasures/treasure_box.png")
        self.image_opened = pygame.image.load("assets/treasures/treasure_box_opened.png")
        self.image = self.image_closed
        self.reward_image_path = reward_image_path  # 報酬アイテム（映画ポスター）
        self.reward_image = pygame.image.load(reward_image_path)  # 報酬アイテム（映画ポスター）
        self.on_ground = False
        self.width = self.image.get_width()
        self.height = self.image.get_height()
        self.is_opened = False
        self.reward_dropped = False  # 報酬がドロップ済みか
        self.blink_counter = 0  # 点滅用カウンタ

    def draw(self, screen, camera):
        """宝箱を描画"""
        if not self.is_opened or (self.is_opened and self.blink_counter % 10 < 5):
            # is_opened かつ 点滅が有効な場合のみ描画
            if self.image:  # self.image が None でないことを確認
                screen.blit(self.image, (self.x - camera.x, self.y - camera.y))

    def update(self, map_instance, clock, TILE_SIZE):
        # 重力と地面判定
        self.on_ground, new_y = map_instance.is_on_ground(self.x, self.y, self.height)
        if self.on_ground:
            self.y = new_y
        else:
            self.y += self.gravity

    def open(self, collected_rewards):
        """宝箱を開ける処理"""
        if not self.is_opened:
            self.is_opened = True
            self.image = self.image_opened
            self.blink_counter = 30  # 点滅用カウンタ初期化
            self.drop_reward(collected_rewards)

    def drop_reward(self, collected_rewards):
        """報酬をドロップ"""
        if not self.reward_dropped:
            self.reward_dropped = True
            collected_rewards.append(self.reward_image)  # アイテムをコレクションに追加

    def handle_event(self, event, collected_rewards):
        """イベント処理"""
        if event.type == pygame.MOUSEBUTTONDOWN:
            mouse_x, mouse_y = event.pos
            if (
                self.x <= mouse_x <= self.x + self.width
                and self.y <= mouse_y <= self.y + self.height
            ):
                self.open(collected_rewards)

    def update_blink(self):
        """点滅演出と宝箱を消す処理"""
        if self.is_opened and self.blink_counter > 0:
            self.blink_counter -= 1